Support queries with no metric name

aaron7 opened the pull request:
Add support for queries with no metric names. e.g. `{foo="bar"}` or `count({job="foo"}) by (__name__)`

Fixes #313 
- Add support to distributor and chunk store
- Create `v7Schema` to support writing metric name entries
- Refactor `IndexEntry` into `IndexEntry` and `IndexQuery` to make separation clearer
- Create a new range value type for metric names and split the code paths so we now support different types of range values
- Change `ExtractMetricNameFromMatchers` to `ExtractMetricNameMatcherFromMatchers` to support optimistations such as matching NotEqual metric name matchers. e.g. `{foo!="bar"}`
- Update bucket functions to return `Bucket`s instead of using callbacks
- Fix bug in hourly bucket where it was offsetting time by day not hour - created a test for this

tomwilkie commented on util/matchers.go:
```diff
@@ -30,19 +30,19 @@ func ExtractMetricNameFromMetric(m model.Metric) (model.LabelValue, error) {
 		}
 		if matcher.Type != metric.Equal {
-			return "", nil, fmt.Errorf("must have equality matcher for MetricNameLabel")
+			return "", nil, false, fmt.Errorf("must have equality matcher for MetricNameLabel")
```
I would remove this error and just return false, as in this case of a non equality matcher (ie a regex, or a !=), we just need to send the request to all ingesters again.

In fact, in the case of non equality, we can do some optimisations on the chunk store too.

tomwilkie commented on distributor/distributor.go:
```diff
@@ -457,21 +457,27 @@ func (d *Distributor) Query(ctx context.Context, from, to model.Time, matchers .
+func (d *Distributor) queryIngesters(ctx context.Context, ingesters []*ring.IngesterDesc, replicationFactor int, req *cortex.QueryRequest) (model.Matrix, error) {
+	// We need a response from a quorum of ingesters, which is n/2 + 1, where n is the replication factor
+	minSuccess := (replicationFactor / 2) + 1
 	maxErrs := len(ingesters) - minSuccess
```
This should be:

```go
maxErrs := replicationFactor / 2
minSuccess :=  len(ingesters) - replicationFactor / 2
```

tomwilkie commented on chunk/chunk_store.go:
```diff
@@ -243,49 +243,104 @@ outer:
+	// Get metric name from matchers
+	metricName, matchers, ok := util.ExtractMetricNameFromMatchers(matchers)
+
+	// If there is a metric name, return chunks using metric name
```
This comment is kinda redundant.

tomwilkie commented on chunk/chunk_store.go:
```diff
@@ -243,49 +243,104 @@ outer:
-	userID, err := user.Extract(ctx)
-	if err != nil {
-		return nil, err
+	// Get metric name from matchers
```
This comment is kinda redundant.

tomwilkie commented on chunk/chunk_store.go:
```diff
@@ -243,49 +243,104 @@ outer:
+	}
+
+	// If there is no metric name, we want return chunks for all metric names
+	metricNameQueries, err := c.schema.GetReadQueries(from, through, userID)
```
need to add:
``go
if err != nil {
    return nil, err
}
```

tomwilkie commented on chunk/chunk_store.go:
```diff
@@ -243,49 +243,104 @@ outer:
+
+	// If there is no metric name, we want return chunks for all metric names
+	metricNameQueries, err := c.schema.GetReadQueries(from, through, userID)
+	metricNameEntries, err := c.lookupEntriesByQueries(ctx, metricNameQueries)
```
need to add:
```go
if err != nil {
    return nil, err
}
```

tomwilkie commented on chunk/chunk_store.go:
```diff
@@ -243,49 +243,104 @@ outer:
+
+	// If there is a metric name, return chunks using metric name
+	if ok {
+		return c.lookupChunksByMetricName(ctx, from, through, matchers, userID, metricName)
```
My preference is to extract the userID from the context at the leaf of the function-tree, as we have to pass contexts for the sake of tracing.

tomwilkie commented on chunk/chunk_store.go:
```diff
@@ -243,49 +243,104 @@ outer:
+
+	incomingChunkSets := make(chan ByKey)
+	incomingErrors := make(chan error)
+	for _, metricNameEntry := range metricNameEntries {
```
Want to add the optimisation here for non-equality metric name matchers - something like:
```go
if !metricNameMatcher.Match(metricNameEntry.Value) {
    continue
}
```

tomwilkie replied:
Will need `ExtractMetricNameFromMatchers` to return the metric name matcher, and not just the name too.

tomwilkie commented on chunk/chunk_store.go:
```diff
@@ -243,49 +243,104 @@ outer:
+	}
+
+	// Merge chunkSets in order
+	return nWayIntersect(chunkSets), lastErr
```
In this case we don't want to interest, but rather union - when doing a query like `{__name__=~"foo.*"}`, we want to return chunks for the both series `foobar` and `foobaz`, not the the intersection of `foobar` and `foobar`  (which will be empty).

tomwilkie commented on chunk/chunk_store.go:
```diff
@@ -243,49 +243,104 @@ outer:
+			return nil, err
+		}
+
+		entries, err := c.lookupEntriesByQueries(ctx, queries)
```
s/lookupEntriesByQueries/lookupEntriesForQueries/

tomwilkie replied:
actually not so sure about this one.

jml replied:
At this point of doing the code review, I'm unsure as to exactly what 'entries' and 'queries' mean, and how they relate to each other. Maybe I'll learn by the time I reach the end.

tomwilkie commented on chunk/chunk_store.go:
```diff
@@ -297,77 +352,85 @@ func (c *Store) lookupMatchers(ctx context.Context, from, through model.Time, ma
-func (c *Store) lookupEntries(ctx context.Context, entries []IndexEntry, matcher *metric.LabelMatcher) (ByKey, error) {
-	incomingChunkSets := make(chan ByKey)
+func (c *Store) lookupEntriesByQueries(ctx context.Context, queries []IndexQuery) ([]IndexEntry, error) {
+	// Call lookupEntriesByQuery for each query
```
Comment is redundant.

tomwilkie commented on chunk/chunk_store.go:
```diff
@@ -376,9 +439,9 @@ func processResponse(ctx context.Context, resp ReadBatch, chunkSet *ByKey, match
+		chunkSet = append(chunkSet, chunk)
 	}
-	return nil
+	return chunkSet, nil
```
Need to sort & dedupe this set for the nWayIntersections above to work.

tomwilkie commented on chunk/inmemory_storage_client.go:
```diff
@@ -138,7 +138,7 @@ func (m *MockStorage) BatchWrite(_ context.Context, batch WriteBatch) error {
 
 // QueryPages implements StorageClient.
-func (m *MockStorage) QueryPages(_ context.Context, entry IndexEntry, callback func(result ReadBatch, lastPage bool) (shouldContinue bool)) error {
+func (m *MockStorage) QueryPages(_ context.Context, entry IndexQuery, callback func(result ReadBatch, lastPage bool) (shouldContinue bool)) error {
```
s/entry/query/

tomwilkie commented on chunk/schema.go:
```diff
@@ -95,193 +97,7 @@ func (cfg *SchemaConfig) RegisterFlags(f *flag.FlagSet) {
 	f.Var(&cfg.V4SchemaFrom, "dynamodb.v4-schema-from", "The date (in the format YYYY-MM-DD) after which we enable v4 schema.")
 	f.Var(&cfg.V5SchemaFrom, "dynamodb.v5-schema-from", "The date (in the format YYYY-MM-DD) after which we enable v5 schema.")
```
Can you move all the SchemaConfig stuff to schema_composite pls?

tomwilkie commented on chunk/chunk_store_test.go:
```diff
@@ -56,6 +56,7 @@ func TestChunkStore(t *testing.T) {
 		{"v4 schema", v4Schema},
 		{"v5 schema", v5Schema},
 		{"v6 schema", v6Schema},
+		{"v7 schema", v7Schema},
```
Can you add a test to the list of cases below for `{__name__=~".+"}`, `{__name__!="foo"}` and `{}` if valid please.

tomwilkie replied:
Oh add something like `{bar=~"beep|baz"}` (without a `__name__` matcher)

tomwilkie replied:
Oh I see you've done it below.  Ignore me!

aaron7 replied:
Yep, I've added below in `TestChunkStoreMetricNames` because the other schemas don't support these queries.

`{}` is not valid on Prometheus - at least one label or a metric name is required.

tomwilkie commented on chunk/inmemory_storage_client.go:
```diff
@@ -125,7 +125,8 @@ func (m *MockStorage) BatchWrite(_ context.Context, batch WriteBatch) error {
 		} else {
-			return fmt.Errorf("Dupe write")
+			// Ignore duplicate write
+			continue
```
Why are we doing duplicate writes?  This will break DynamoDB if we do a dupe write in the same batch.  Need to put the error back in, and add this as a comment.

aaron7 replied:
The reason duplicate writes are now appearing is because of the metric name `IndexEntry`. Multiple chunks have the same `bucketHashKey` and the `RangeValue` doesn't include the `chunkID`. 

I will add back the error and dedupe the writes.

aaron7 replied:
One solution would be to move the metric name entry from `GetWriteEntries` to something like `GetMetricNameWriteEntries` and then only call it in `calculateDynamoWrites` if we haven't seen the metric name before.

jml replied:
@aaron7 Is this what we spoke about the other day? What's happened since then?

tomwilkie commented on chunk/schema.go:
```diff
@@ -343,46 +118,101 @@ func v6Schema(cfg SchemaConfig) Schema {
+	var result []IndexEntry
+
+	buckets := s.buckets(from, through, userID)
+	for _, bucket := range buckets {
```
Nit, but IMO this could be `for _, bucket := range s.buckets(from, through, userID) {`

tomwilkie commented on chunk/schema.go:
```diff
@@ -343,46 +118,101 @@ func v6Schema(cfg SchemaConfig) Schema {
+
+	buckets := s.buckets(from, through, userID)
+	for _, bucket := range buckets {
+		entries, err := s.entries.GetWriteEntries(bucket.from, bucket.through, bucket.tableName, bucket.hashKey, metricName, labels, chunkID)
```
You should pass `bucket` directly to `GetWriteEntries`:
```go
entries, err := s.entries.GetWriteEntries(bucket, metricName, labels, chunkID)
```
I always think too many arguments is a bit of a small.

jml replied:
"Too many" of anything is always bad—that's what "too many" means.

tomwilkie commented on chunk/schema.go:
```diff
@@ -394,41 +224,45 @@ func (originalEntries) GetWriteEntries(_, _ uint32, tableName, hashKey string, l
-func (originalEntries) GetReadMetricEntries(_, _ uint32, tableName, hashKey string) ([]IndexEntry, error) {
-	return []IndexEntry{
+func (originalEntries) GetReadQueries(_, _ uint32, _, _ string) ([]IndexQuery, error) {
+	return nil, fmt.Errorf("originalEntries does not support GetReadQueries")
```
Its probably time to make an ErrNoMetricName constant at the top of the file and return that.  Plus, lets make the error message something a user can understand "metric name required for pre-v7 schemas" or something.

tomwilkie commented on chunk/schema.go:
```diff
@@ -561,37 +403,41 @@ func (v5Entries) GetWriteEntries(_, through uint32, tableName, hashKey string, l
 			TableName:  tableName,
-			HashValue:  hashKey + ":" + string(key),
-			RangeValue: buildRangeKey(encodedThroughBytes, encodedValueBytes, chunkIDBytes, rangeKeyV4),
+			HashValue:  bucketHashKey + ":" + string(metricName) + ":" + string(key),
```
I've always personally preferred `fmt.Sprintf` for this kind of thing, when there are more than one or two `+`s.

jml replied:
Not sure how Go implements under the hood, but in Python or Java, sprintf would be more performant, as repeated string concatenation is quadratic.

tomwilkie commented on chunk/schema.go:
```diff
@@ -618,128 +464,105 @@ func (v6Entries) GetWriteEntries(_, through uint32, tableName, hashKey string, l
-	return encoded
-}
+func (v7Entries) GetWriteEntries(_, through uint32, tableName, bucketHashKey string, metricName model.LabelValue, labels model.Metric, chunkID string) ([]IndexEntry, error) {
+	metricName, err := util.ExtractMetricNameFromMetric(labels)
```
Please call `entries.v6Entries.GetWriteEntries(...)` then add the extra entry to that, should reduce code duplication.

tomwilkie commented on chunk/schema_util.go:
```diff
@@ -0,0 +1,116 @@
+	return output
+}
+
+func deconstructRangeKey(value []byte) [][]byte {
```
👍 but can we either call `buildRangeKey` `constructRangeKey` or change this to `unbuildRangeKey`?  Or perhaps call `buildRangeKey` `mantleRangeKey` and make this one `dismantleRangeKey`?

jml replied:
Other classic variants: `toRangeKey` & `fromRangeKey`; `encodeRangeKey` & `decodeRangeKey`; `unparse` and `parse` .

-1 on `mantle`.

tomwilkie commented on util/matchers.go:
```diff
@@ -10,7 +10,8 @@ import (
 func SplitFiltersAndMatchers(allMatchers []*metric.LabelMatcher) (filters, matchers []*metric.LabelMatcher) {
 	for _, matcher := range allMatchers {
-		if matcher.Match("") {
+		// Test for empty filters. Other metrics such as NotEqual are not filters.
```
Why not?

aaron7 replied:
As mentioned in the original comment: "SplitFiltersAndMatchers splits empty matchers off, which are treated as filters, see #220"

Unfortunately `{foo!="bar"}` matches `""`, which means it also gets split off as a filter and we don't get a chance to apply optimisations before we return any chunks. I came across this with queries such as `{__name__!="foo"}` and with our optimisation before fetching chunks not being made because the matcher is filtered out from matchers.

Is this correct? I can add some more comments explaining if it is.

aaron7 replied:
I've gone through this again and understand why we are filtering out all matchers matching "". I will revert the change and describe in a comment why.

tomwilkie commented:
Looking really good @aaron7!  Probably needs one more round of review after you've addressed these minor comments.  Have you tested this locally, on minikube with the local dynamodb and with the v7 schema enabled?

Also, time to squash it all down and rebase I think.

jml commented on chunk/by_key_test.go:
```diff
@@ -9,7 +9,70 @@ func c(id string) Chunk {
+		have := nWayMerge(tc.in)
+		if !reflect.DeepEqual(have, tc.want) {
+			t.Errorf("%v != %v", have, tc.want)
+		}
```
`assert.Equal` here as well please.

jml commented on chunk/by_key_test.go:
```diff
@@ -9,7 +9,70 @@ func c(id string) Chunk {
+			t.Errorf("%v != %v", have, tc.want)
+		}
+	}
+}
```
This is a really minor point, but your `nWayMerge` algorithm assumes `ByKey` and `merge` form a monoid. You should probably have a test for associativity and identity properties. (I think the code works and the test coverage is enough, so no need to actually _do_ this).

jml commented on chunk/chunk_store.go:
```diff
@@ -243,49 +243,128 @@ outer:
+		}
+
+		// If there is a metric name matcher, however we are fetching all metric name
+		// chunks, skip metric names which don't match the metric name matcher.
```
This doesn't make sense grammatically. I *think* it means…

```golang
// We are fetching all metric name chunks, but we only want metric names that match metricNameMatcher.
```

Is that right?

aaron7 replied:
Yep, I've updated the comment to make it clearer.

jml commented on chunk/chunk_store.go:
```diff
@@ -297,77 +376,84 @@ func (c *Store) lookupMatchers(ctx context.Context, from, through model.Time, ma
 		}
 	}
 
+	// Merge chunkSets in order
```
Not sure this comment is helpful. More interesting to me would be a comment saying why this is an intersect and not a merge.

jml commented on chunk/by_key.go:
```diff
@@ -55,6 +56,27 @@ func merge(a, b ByKey) ByKey {
 
+// nWayMerge will merge and dedupe n lists of chunks.
+// lists must be sorted and not contain dupes.
+func nWayMerge(sets []ByKey) ByKey {
```
Maybe `nWayUnion` to parallel `nWayIntersect`?

jml commented on chunk/schema_config.go:
```diff
@@ -0,0 +1,328 @@
+	return cfg.TablePrefix + strconv.Itoa(int(bucketStart/int64(cfg.TablePeriod/time.Second)))
+}
+
+// Bucket desribes a range of time with a tableName and hashKey
```
Describes.

aaron7 commented:
@jml, @tomwilkie - thanks for the reviews. I've moved the `schema.go` split into another PR and rebased this on top of it.

The only thing left for this diff is discussing DynamoDB writes using `BatchWriteItem` with the duplicate metric name index entries.

jml commented:
Lovely, thanks. Let's talk about it Tuesday morning!

jml commented:
Sorry @aaron7, I've kind of lost track of this. Are you waiting for another round of review?

aaron7 commented:
@jml would it be possible to review the last commit? that is all that is not reviewed yet :)

jml commented on chunk/chunk_store.go:
```diff
@@ -161,6 +161,8 @@ func (c *Store) updateIndex(ctx context.Context, userID string, chunks []Chunk)
 // calculateDynamoWrites creates a set of batched WriteRequests to dynamo for all
 // the chunks it is given.
 func (c *Store) calculateDynamoWrites(userID string, chunks []Chunk) (WriteBatch, error) {
+	seenIndexEntries := map[string]bool{}
```
Another way of doing this is `map[string]struct{}{}` i.e. a map from strings to empty structs. No need to change this though, `bool` is also fine.

jml commented on chunk/chunk_store.go:
```diff
@@ -174,7 +176,17 @@ func (c *Store) calculateDynamoWrites(userID string, chunks []Chunk) (WriteBatch
 		indexEntriesPerChunk.Observe(float64(len(entries)))
 
+		// Remove duplicate entries based on tableName:hashValue:rangeValue
+		filteredEntries := []IndexEntry{}
```
`unseenEntries` probably communicates intent a bit more.

tomwilkie commented:
I'm going to do a bit of manual testing locally now...

tomwilkie commented:
Have brought this up using the yamls in k8s/ and tested the following query: http://nginx.default.svc.cluster.local/api/prom/api/v1/query?query=count(%7B__name__%3D~%22.%2B%22%7D)+by+(__name__)

Seems to work!  This will have mostly just hit the ingesters though, so I'm going to do more testing to see what happens with the chunk store.

tomwilkie commented:
I get the following error: `Error in MergeQuerier.Query: unrecognised metricNameRangeKey version: ''`

tomwilkie commented:
Seems like an extra nil byte is getting added at the front `unrecognised metricNameRangeKey version: '[[] [12 43 245 52 209 113 74 187 49 166 49 159 65 135 83 147 20 46 214] [] [] [54]]'`

tomwilkie commented:
Ahh - the has could contain a nil byte.  need to base 64 encode it.

tomwilkie commented:
With that change I no longer get the error, but it is much slower than I expected.  I suspect this is just the local dynamodb being slow, but we should see how it performs in dev.

All my testing has been on the v7 schema only, so when this lands in dev (with the v6 schema refactorings) we need to check how it behaves.
